hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerChecksumTreeManager.java

```diff
@@ -143,12 +149,148 @@ public void markBlocksAsDeleted(KeyValueContainerData data, Collection<Long> del
     }
   }
 
-  public ContainerDiff diff(KeyValueContainerData thisContainer, ContainerProtos.ContainerChecksumInfo otherInfo)
-      throws IOException {
-    // TODO HDDS-10928 compare the checksum info of the two containers and return a summary.
-    //  Callers can act on this summary to repair their container replica using the peer's replica.
-    //  This method will use the read lock, which is unused in the current implementation.
-    return new ContainerDiff();
+  public ContainerDiff diff(KeyValueContainerData thisContainer, ContainerProtos.ContainerChecksumInfo peerChecksumInfo)
+      throws Exception {
+    ContainerDiff report = new ContainerDiff();
+    try {
+      captureLatencyNs(metrics.getMerkleTreeDiffLatencyNS(), () -> {
+        Preconditions.assertNotNull(thisContainer, "Container data is null");
+        Preconditions.assertNotNull(peerChecksumInfo, "Peer checksum info is null");
+        Optional<ContainerProtos.ContainerChecksumInfo.Builder> thisContainerChecksumInfoBuilder =
+            read(thisContainer);
+        if (!thisContainerChecksumInfoBuilder.isPresent()) {
+          throw new IOException("The container #" + thisContainer.getContainerID() +
+              " doesn't have container checksum");
+        }
+
+        if (thisContainer.getContainerID() != peerChecksumInfo.getContainerID()) {
+          throw new IOException("Container Id does not match for container " + thisContainer.getContainerID());
```

Comment:
Should we preserve the fidelity of the two different scenarios in which this method raises an exception? If the file does not have any content. Also. `read` can throw an exception. The container ID not matching is a more catastrophic error as that would be a coding bug vs issues with filesystem read.

Reply:
Handled in the latest change.